File: backend/app/schemas/stats.py
from pydantic import BaseModel
from datetime import date
from typing import Optional


class StatsResponse(BaseModel):
    total_xp: int
    level: int
    xp_for_next_level: int
    xp_progress_in_level: int
    quizzes_completed: int
    streak_days: int
    best_streak: int
    average_score: float
    last_active: Optional[date]


class SubjectStat(BaseModel):
    document_title: str
    attempts: int
    best_score: float
    average_score: float
    total_xp: int


class SubjectStatsResponse(BaseModel):
    subjects: list[SubjectStat]


class ScoreHistoryItem(BaseModel):
    date: str
    score: float
    document_title: str


class ScoreHistoryResponse(BaseModel):
    history: list[ScoreHistoryItem]

File: frontend/MarkedQuiz/Models/Stats.swift
import Foundation

struct StatsResponse: Codable, Sendable {
    let totalXp: Int
    let level: Int
    let xpForNextLevel: Int
    let xpProgressInLevel: Int
    let quizzesCompleted: Int
    let streakDays: Int
    let bestStreak: Int
    let averageScore: Double
    let lastActive: String?

    enum CodingKeys: String, CodingKey {
        case level
        case totalXp = "total_xp"
        case xpForNextLevel = "xp_for_next_level"
        case xpProgressInLevel = "xp_progress_in_level"
        case quizzesCompleted = "quizzes_completed"
        case streakDays = "streak_days"
        case bestStreak = "best_streak"
        case averageScore = "average_score"
        case lastActive = "last_active"
    }
}

struct SubjectStat: Codable, Identifiable, Sendable {
    let documentTitle: String
    let attempts: Int
    let bestScore: Double
    let averageScore: Double
    let totalXp: Int

    var id: String { documentTitle }

    enum CodingKeys: String, CodingKey {
        case attempts
        case documentTitle = "document_title"
        case bestScore = "best_score"
        case averageScore = "average_score"
        case totalXp = "total_xp"
    }
}

struct SubjectStatsResponse: Codable, Sendable {
    let subjects: [SubjectStat]
}

struct ScoreHistoryItem: Codable, Identifiable, Sendable {
    let date: String
    let score: Double
    let documentTitle: String

    var id: String { "\(date)-\(documentTitle)-\(score)" }

    enum CodingKeys: String, CodingKey {
        case date, score
        case documentTitle = "document_title"
    }
}

struct ScoreHistoryResponse: Codable, Sendable {
    let history: [ScoreHistoryItem]
}
